Sources/Target/POSIX/ELFProcess.cpp

```diff
@@ -53,52 +53,66 @@ inline ErrorCode GetELFSharedLibraryInfoAddress(ELFProcess *process,
+  uint64_t phdrAddress = process->getAuxiliaryVectorValue(AT_PHDR);
+  size_t phdrEntrySize = process->getAuxiliaryVectorValue(AT_PHENT);
+  size_t phdrNumEntry = process->getAuxiliaryVectorValue(AT_PHNUM);
+  uint64_t loadBase = phdrAddress - sizeof(eh);
+  if (phdrAddress == 0 || phdrEntrySize == 0 || phdrNumEntry == 0) {
+    DS2LOG(Debug, "Unable to locate PHDR info.");
+    DS2LOG(Debug, "Dynamic Library info will not be available.");
     return kErrorUnsupported;
   }
 
-  // 1. Find PT_DYNAMIC (.dynamic) program header, and the
-  //    expected load base of this ELF.
+  // 2. Find PT_DYNAMIC (.dynamic) program header and what the ELF believes is
+  //    the loadBase. Find the size of the dynamic section as well.
   uint64_t elfLoadBase = 0;
   bool foundElfLoadBase = false;
-  uint64_t phdrAddress = process->loadBase() + eh.e_phoff;
-  for (size_t n = 0; n < eh.e_phnum; n++) {
-    CHK(process->readMemory(phdrAddress, &ph, sizeof(ph)));
-
-    if (!foundElfLoadBase && ph.p_type == PT_LOAD) {
-      elfLoadBase = ph.p_paddr;
+  uint64_t dotDynamicAddress = 0;
+  bool foundDotDynamicAddress = false;
+  uint64_t dotDynamicMemsz = 0;
+  for (size_t n = 0; n < phdrNumEntry; n++) {
+    CHK(process->readMemory(phdrAddress + n * phdrEntrySize, &ph, sizeof(ph)));
+
+    if ((!foundElfLoadBase || ph.p_vaddr < elfLoadBase) &&
+        ph.p_type == PT_LOAD) {
+      elfLoadBase = ph.p_vaddr;
       foundElfLoadBase = true;
     }
 
     if (ph.p_type == PT_DYNAMIC) {
-      break;
+      dotDynamicAddress = ph.p_vaddr;
+      dotDynamicMemsz = ph.p_memsz;
+      foundDotDynamicAddress = true;
     }
-
-    phdrAddress += eh.e_phentsize;
   }
 
-  if (ph.p_type != PT_DYNAMIC) {
+  if (!foundDotDynamicAddress || !foundElfLoadBase) {
+    DS2LOG(Debug, "Unable to locate PT_DYNAMIC or any PT_LOAD.");
+    DS2LOG(Debug, "Dynamic Library info will not be available.");
     return kErrorUnsupported;
   }
 
-  // 2. Find DT_DEBUG in .dynamic section
-  uint64_t dynsymAddress = process->loadBase() + ph.p_paddr - elfLoadBase;
-  size_t entriesCount = ph.p_memsz / sizeof(dynsym);
+  // 3. Find DT_DEBUG in .dynamic section
+  dotDynamicAddress += (loadBase - elfLoadBase);
+  size_t entriesCount = dotDynamicMemsz / sizeof(dynsym);
   for (size_t n = 0; n < entriesCount; n++) {
-    CHK(process->readMemory(dynsymAddress, &dynsym, sizeof(dynsym)));
+    CHK(process->readMemory(dotDynamicAddress + n * sizeof(dynsym), &dynsym, sizeof(dynsym)));
 
-    // 3. Return &DT_DEBUG->d_ptr, which is filled by the dynamic linker
+    // 4. Return &DT_DEBUG->d_ptr, which is filled by the dynamic linker
     //    pointing to the link_map.
     if (dynsym.d_tag == DT_DEBUG) {
-      address = dynsymAddress + sizeof(dynsym.d_tag);
+      address = (dotDynamicAddress + n * sizeof(dynsym)) + sizeof(dynsym.d_tag);
       return kSuccess;
     }
 
-    dynsymAddress += sizeof(dynsym);
   }
 
+  DS2LOG(Debug, "DT_DEBUG was not found in .dynamic");
```

Comment:
Same here, just one log message is better.

    DS2LOG(Debug, "DT_DEBUG was not found in .dynamic, "
                  "dynamic Library info will not be available.");